File: src/main/java/editor/serialization/VersionAdapter.java
package editor.serialization;

import java.lang.reflect.Type;
import java.text.ParseException;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import editor.database.version.DatabaseVersion;

public class VersionAdapter implements JsonSerializer<DatabaseVersion>, JsonDeserializer<DatabaseVersion>
{
    @Override
    public JsonElement serialize(DatabaseVersion src, Type typeOfSrc, JsonSerializationContext context)
    {
        return new JsonPrimitive(src.toString());
    }

    @Override
    public DatabaseVersion deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
    {
        try
        {
            return new DatabaseVersion(json.getAsString());
        }
        catch (ParseException e)
        {
            throw new JsonParseException(e);
        }
    }
}

File: src/main/scala/editor/serialization/VersionAdapter.scala
package editor.serialization

import editor.database.version.DatabaseVersion
import org.json4s.CustomSerializer
import org.json4s.JString

/**
 * JSON serializer/deserializer for [[DatabaseVersion]]s using the format specified by [[DatabaseVersion.toString]].
 * @author Alec Roelke
 */
class VersionAdapter extends CustomSerializer[DatabaseVersion](formats => (
  { case JString(version) => DatabaseVersion.parseVersion(version) },
  { case version: DatabaseVersion => JString(version.toString) }
))